exercises/nucleotide-count/canonical-data.json

```diff
@@ -1,12 +1,12 @@
 {
   "exercise": "nucleotide-count",
-  "version": "1.0.0",
+  "version": "1.1.0",
   "cases": [
     {
-      "description": "count all nucleotides in a strand",
+      "description": "nucleotideCounts counts all DNA-nucleotides in the input strand and returns an object with properties A, C, G and T holding their respective amounts.",
       "cases": [
         {
-          "description": "empty strand",
+          "description": "counts an empty DNA strand as 0.",
```

Comment:
The original name here was fine, you are adding redundant information here.

Reply:
You are right. Though, my intent was to implement well-formed descriptions in the informal style of 
´´´{it / the test} does xyz in the case of abc.```
What I've experienced so far, this is not even enough. A test case (in TDD especially) acts as a specification. There are best practices (in some x-driven-development movements even rules) that suggest a test description should have several parts:
- in case of x
- then y
- reasoning
- some also want a precondition (like BDDs GIVEN WHEN THEN)

Just stating "empty string" as description tells a developer absolutely nothing about what should happen (or even why). I know you care about the reasoning part the same way I do, so I am not sure why you want to omit it here.

So in my eyes my suggested description might even be phrased like:
"Given an empty input string, it returns zeroes for all nucleotides."
While the reasoning for this behaviour results from the description of the test suite, because there is no need to state this dogmatically in every description.

Reply:
> "counts an empty DNA strand as 0."

'counts an' is redundant because all function calls will "count"

'DNA' is redundant because all input is DNA strands.

'as 0.' is information already present in the expected output.

Which leaves us with "empty strand"

Which means: "Returns the correct value when provided with an empty strand as input"

> my intent was to implement well-formed descriptions in the informal style of
{it / the test} does xyz in the case of abc.

If a description can be generated from the data within the test it adds no value.
A descriptions should describe the "why" of the test. What does this test test that all the others do not?

Reply:
I see it this way:
- When running a test suite, you don't see the data, but only the names. So if for example something fails, then the test name itself should already give a clue about what might be the error or the constraints of the error. 
- When a user implements code inside an exercise track from a test spec or does bug fixing, he/she sees the data of the spec. But this only helps with trivial test logic. Say you have some more complex test logic, even though its intention can be described rather specifically in a sentence. Then why not being specific in the first place?
- When a track maintainer implements a test from the canonical data, he/she sees the name and the data. But to create a meaningful test case he/she has to be able to know exactly what the test is about. Inferring things implicitly leads more often than not to misunderstandings. ("Oh, you meant this? I thought you meat that?") I think every programmer knows what it means when the specifications are unclear or when the customer has a different understanding than the developers. That's the reason why whole methodologies like Domain Driven Design evolved in the first place. I can't think of any requirement engineer who ever said: "Meh, no need to be redundant with this spec, this is obvious."

When you just say "empty strand", noone could infer 
> "Returns the correct value when provided with an empty strand as input"
without context. It might as well mean that an empty strand should be returned. 
Yes, in this particular context here it might be obvious, but e.g. in a bit more complex production code it ain't. 

So the question that comes into my mind is: 
Why is redundancy here in the canonical data considered a bad thing?
Don't get me wrong, being concise is nice and logical and good practice indeed. But is it pragmatic from the users or implementers perspective? Is it readable and most importantly understandable and unambiguous? Does a beginning learner already has the knowledge/skill to infer the meaning from the context?
For example: I just recently saw someone learning the very basics of programming from a nicely done beginners website. And he struggled with the easiest and basic tasks where everyone else with a bit more understanding would knew what was going on in an instant. But he was not fully aware of the context that was given and therefore could not infer it easily.

So the TL;DR ^^:
Why do you consider redundancy in a spec file (more so even meta-spec file) a bad thing if it helps making the matter more clear and unambiguous?

Reply:
> When running a test suite, you don't see the data, but only the names.

You need a better test runner. I see the expected input and output values and highlighted diffs of the results.

>  this only helps with trivial test logic. Say you have some more complex test logic

I'm not sure what your point is in this paragraph, but if your test logic is complex you're doing something wrong.


>  to create a meaningful test case he/she has to be able to know exactly what the test is about

Several tracks implement test generators that automatically generate the tests without the track maintainer needing to look at any of the tests, so I would argue this statement is false.

> Why do you consider redundancy in a spec file (more so even meta-spec file) a bad thing if it helps making the matter more clear and unambiguous?

Redundancy in a description is often a sign that the description is not telling me why this test is important, but is just repeating what the inputs and outputs are.

If it really is making things clear an unambiguous is it really redundant?
(see: no true Scotsman)